### ISSUE TITLE
Update BRouter Android app

### ISSUE THREAD

**Issue by afischerdev** [abrensch/brouter#312]
I'd like to update the app to Android 11

These are the plans:

Starting with switch to gradle, issue [243](https://github.com/abrensch/brouter/pull/243)
This has no conflicts at the moment
 
Android 11
A good start is 'Cleanup file / base directory access' issue [244 ](https://github.com/abrensch/brouter/pull/244)
but it has conflicts 

The new storage rule says use your own space to have free file access or work with the new DocumentFile class.
But own space in .../Android/data/btools.routing/ is no longer visible for others.
And we can't see data space from other apps as well.

One way to live with that rules seems the use of the media path
.../Android/media/btools.routing/
Here we have random file access and it is visible for other apps.

That also means the rules for the CoordinateReader should change.
E.g. a folder '/Android/media/btoos.routing/income' could be open for all apps to write there gpx into it.
No grep on special app folders as before.

New app icon - brouter-web issue [286](https://github.com/nrenner/brouter-web/issues/286)
I don't know if we follow here or use a Android specific way
see [Android specs](https://developer.android.com/google-play/resources/icon-design-specifications)

Issue #246
Export file or direct output json 

Issue #306
Update all tiles

Switch to server logic input:
- parameter coordinate to 'lonlats', keep old parameter
- enable nogos, polylines, polygons
- enable pois
- other parameters

Update lookup table logic
One of the reason the Android app is outdated is the fixed lookup table.
What about an update function based on the lookup version tags?
E.g. when download rd5 file and it has a new version number, load also a new lookups.dat.
As well as new profiles.
Or: an update file that contains the updatable profiles,
e.g. we have a new tag 'highway=electro-only', we only need to update the car profile.
Or: update on different file size

Issue #301
foot maxspeed hardcoded
Not a direct Android app issue.
It affects the StdPath class which enables speed handling only for foot or bike. I'd like to enable others as well.

Issue #286
Black screen when swiping up or down

Issue #277
brouter-service and Android watchdog
Don't know what to do here but I'll have an eye on it. I know the Android is killing services but never had this issue with BRouter. 

Issue #271
Download manager
I can't follow this error message on my Android 11 hardware but may be @gammeredith can help with testing a new version.

Issue #227
Download Manager - not running in background
A good idea to get a handy app.

Maybe I've forgotten something that could change.

**Comment by polyscias** [comment]
I saw the email coming by on the email-list and did not come yet to react but let me do it here.

This looks like a nice list to me and would love if it would be implemented

**Comment by EssBee59** [comment]
Hello Afischerdev,
Very good  idea to create a new version!

Here some remarks:

A-To my issue 277: I think, the problem is a result of the Android behaviour with processes running in backgroud (IPC) that are killed according to  (for me) unknown creterias/ after an undefined time...
I could solve it with a patch in Osmand (reconnect in case of error), but the Osmand team had other plans with Brouter...
I did not found a solution to avoid the "process kill" in Brouter itself, so callers (as ex Osmand) should enhance their resiliance!

B- Storage in Android 11
I placed the Brouter in .../brouter  (root of the data directory) and it also works without  problems
(Perhaps is your  .../Android/media/btools.routing/ better, I did not test it)
See also https://github.com/osmandapp/OsmAnd/issues/10453

C- Nogo
Do you have plans with nogos?
Currently 2 versions are available (API11 and API28) 
On Android 11 I installed Osmand on .../Osmand and can read the favourites.gpx with the api11 version.
But for future versions I think the interface osmand-brouter should be extended (as ex. with the nogo data and the name of the profile to be used)

Regards
Ess Bee

**Comment by afischerdev** [comment]
Hello Ess Bee,

thanks for your comments.

> B- Storage in Android 11
I placed the Brouter in .../brouter (root of the data directory) and it also works without problems
(Perhaps is your .../Android/media/btools.routing/ better, I did not test it)

Interesting thing, I didn't get it installed on the root - after select an existing folder. Did you come from Android 10 to 11 with this folder? I own a Pixel that was delivered with Android 11 and I can only install on ../Android/data/..

> C- Nogo
Do you have plans with nogos?
Currently 2 versions are available (API11 and API28)
On Android 11 I installed Osmand on .../Osmand and can read the favourites.gpx with the api11 version.
But for future versions I think the interface osmand-brouter should be extended (as ex. with the nogo data and the name of the profile to be used)


My favorite is an incoming folder in a 'public' folder where BRouter can read with file access and not via DocumentFile. In other context I tested reading mapsforge layouts from several place and I had to learn that reading content via DocumentFile and DocumentTree is very slow - sometimes factor is more than 10. See also [here ](https://stackoverflow.com/questions/64842185/android-11-documentfile-findfile-is-too-slow) or similar.
But an extra way for the interface is also possible. The server already provides parameter 'nogos', 'polylines' or 'polygons'.
I think it belongs to the size of data we have to handle.

All this will produce some work for other apps.

**Comment by afischerdev** [comment]
@polyscias 

> This looks like a nice list to me and would love if it would be implemented

Feel free to add other useful things for the app.

**Comment by EssBee59** [comment]
> Interesting thing, I didn't get it installed on the root - after select an existing folder. Did you come from Android 10 to 11 with this folder? I own a Pixel that was delivered with Android 11 and I can only install on ../Android/data/..

Yes, my smartphone was migrated from Android 10 to Android 11. But I made a new install from brouter to the new location.
I am using the "FX" file manager on my Xiaomi smartphone, can create directories on the root without problems.

The idea with a new interface is, if possible to get rid of any "storage access rights", it would make a distribution over google play store much easier, and simplify the configuration in brouter (location of Osmand as ex.)
Contact with the other apps developpers (Osmand..) would be necessary to go in that direction.

My dream is to implement also in the interface the transfer of the profile-name  that is to be used:
I created own versions of Brouter and Osmand supporting that, then I created in Osmand different user-profiles with each different names for the brouter-profiles.
So now I can in Osmand select by starting a new tour the profile I need (MTB, Fastbike, Trekking)...

**Comment by afischerdev** [comment]
> My dream is to implement also in the interface the transfer of the profile-name that is to be used:
I created own versions of Brouter and Osmand supporting that, then I created in Osmand different user-profiles with each different names for the brouter-profiles.
So now I can in Osmand select by starting a new tour the profile I need (MTB, Fastbike, Trekking)...

I've added a new issue #314 for the interface features to make it clearer.

**Comment by EssBee59** [comment]
Hello,
I come back to the basic problem of storage access under android 11:
The different options:

Option A- Brouter data stored in .../Android/data/btools.routingapp/files/brouter
Advantage:  the brouter-app do not need "storage access" rights
Disadvantage: manual access to the "brouter.data" is not easy but possible (as example using the "Explorer" app of "Speed Software" or accessing from a PC via USB)

Option B- Brouter data stored in  .../Android/media/btools.routingapp/files/brouter or .../brouter
Advantage:   access to the "brouter.data" is further possible as under android 10
Disadvantage:  the brouter-app NEEDS "storage access" rights

Are my assumptions and tests right?
To avoid stress with google play store, I think (in long term) option A should be better, but as explained the old api11 version neede storage access rights for nogo and WP´s.

**Comment by afischerdev** [comment]
@EssBee59 
> Option B- Brouter data stored in .../Android/media/btools.routingapp/files/brouter or .../brouter
> Advantage: access to the "brouter.data" is further possible as under android 10
> Disadvantage: the brouter-app NEEDS "storage access" rights

B is not correct. switch to media path does not need a permission. 
I did a test project for this feature, please have a look to [this ](https://github.com/afischerdev/brouter/tree/test-and11)
It only changes the basedir. You could also remove the WRITE_EXTERNAL_STORAGE permission. Will still work.
brouter folder inn main still does not work for me.

**Comment by EssBee59** [comment]
> B is not correct. switch to media path does not need a permission.

Yes, I retested with in .../Android/media/btools.routingapp/files/brouter on my Xiaomi/Android 11, no read permission required!
BUT, I had to create MANUALLY the directories "btools.routingapp" and "files" in .../Android/media with a file manager having storage permission.
Could you install direct with brouter only?

As I can install on .../, we have possibly different Androids and behaviours??

**Comment by afischerdev** [comment]
@EssBee59 
>Could you install direct with brouter only?

Yes, the test app does all for me.

>As I can install on .../, we have possibly different Androids and behaviours??

Can't say that. BRouter test app uses targetSdkVersion 30 now

**Comment by EssBee59** [comment]
If you can send me your "sdk 30" test version, I could test it on my Xiaomi !!!
(I am not sure to be able to generate such a version quickly)

**Comment by afischerdev** [comment]
@EssBee59 
Ok, [here is comes](https://www.dropbox.com/s/qw20h2guqo4c0mb/BRouterApp.1.6.1-test-release.apk?dl=0). It's a dropbox link (you don't need dropbox to install). When like to do install direct to Android you please only download it, not store in your dropbox, installation from there doesn't work anymore. 
And keep in mind, it's a version signed with my values, so you have to uninstall other version before.

**Comment by EssBee59** [comment]
Result of my retest:
- direct installation in ../Android/media/btools.routingapp  is OK (without storage access permission)
- direct installation in ../   is not OK (also with storage access permission)
So far, same as your results.
Not a real surprise: "target sdk 30" is prerequisited for optimal use under Android 11 !!!!!!!

**Comment by afischerdev** [comment]
@EssBee59 
Thanks for the test reporting.
New Apps muss target to this from August 21, old app from Nov. 21 [(see Google Timeline)](https://developer.android.com/distribute/play-policies#APILevel30)

**Comment by afischerdev** [comment]
Update for the app seems to be nearly ready.

still pending
- git workflow yml script for gradle
- Network connection issue #321 #322 should be closed when no reply
- refreshed profiles
  we have different brf in misc/profiles and brouter.de/brouter/profiles2/
  So I would like to normalize and use the brouter.de brfs and use this as update pool as well
  And shouldn't we establish an automatic process on profiles to export git to brouter.de/brouter/profiles2/ ? May be with git workflow?
  
- brouter-service and Android watchdog #277 - still no error results during testing download manager
- Download manager. #271 - the download logic is new, so I think we can close


others
- I added an icon to start discussion again, thinking of ways, nodes and mountains, no comments until now. [view icon](https://github.com/abrensch/brouter/blob/master/brouter-routing-app/src/main/res/mipmap-hdpi/ic_launcher_round.png)
- BInstallerSizes: could someone generate a new version of BInstallerSizes? Or just place a new factor?
- what about adding issue feature request: waterway routing #233 ?
  Headline is waterway routing, but it has in general the possibility to add variable values to the lookups.dat and does not affect the current lookup. Contains converter for feet, fathom, miles, mph, knots, like
`  maxheight;0000000001 *`
- I would like to add the pull request #289 - didn't notice it before

last actions 
- new version number 
- for the Play Store a bundle *.aab file is needed for publishing
  This could be generated with
  './gradlew :brouter-routing-app:bundleRelease'
  But this will need a private_key.pepk file to do the signing in the Play Store. Google offers a signing from Google but then the app become a new key and is not compatible with the previous key version.
  I didn't found a way to this with gradle. But this can be done with the Android Studio.
  So I think this is a job for the owner of the original key.
  F-Droid can generate there own apk with gradle.
  And hardcore user can install the app from the brouter....zip file.

**Comment by EssBee59** [comment]
Hi !
fantastic job!
Just a short comment to issue I created last year "brouter-service and Android watchdog #277":
I am still using intensively OSMAND + Brouter, but I did not observed the problem again in the last weeks!!!

(a lot of updates were installed on my Xiaomi in the last monthes, as example migration to Android 11 and  MIUI 12.5)

So, if nobody else reports similar problems, I suggest to close the issue now (by need, I can reopen or create a new one)

regards
Ess Bee

**Comment by afischerdev** [comment]
@EssBee59 
Fine, great, I agree fully.

**Comment by abrensch** [comment]
>> for the Play Store a bundle *.aab file is needed for publishing

what I did so far was just upload an apk, build with the path to the certificate placed somewhere in the maven-pom

I can give it a try with the gradle-build. Or just send you the certificate by email?

Is it still useful to build the "Api10" version for direct installs without security restrictions? Or is this all history?

thanx for your great work, Arndt

**Comment by afischerdev** [comment]
Hi Arndt,

> what I did so far was just upload an apk, build with the path to the certificate placed somewhere in the maven-pom

I know, was the same with my apps.
But now a bundle is required.
more info about bundle at [Android blog](https://medium.com/androiddevelopers/building-your-first-app-bundle-bbcd228bf631)
I didn't get the jarsigner run with my parameters to build a signed aab.  We could try it with Android Studio with your mykey.store file. So if you like to do it that way please send off topic the keystore file and the passphrase1 and 2 and the alias. I would send the signed *.aab and the private_key.pepk which is needed for upload on Play Store for Google to resign all build variantes.

> Is it still useful to build the "Api10" version

I can't say that. Current version is backwards usable with Android 19 (4.4 2013, Android 10 is 2.3.3 from 2011). I think older libraries are needed for compilation.  I could give it a try.

**Comment by EssBee59** [comment]
API10 vs API28 versions
Hello  afischerdev,
I think, Arndt created these different versions to support „nogo“ areas and avoid conflicts with google policies :

API10 : (manual installation)  « storage access » is required for the app. So, brouter can read (as example by using with OSMAND) the file « favourites.gpx » in the root directory of the OSMAND app.
(the nogo-list is then created with this file)

API28 : (Google installation) no « storage access » is granted at startup, « hard coded », brouter do not try to read the favourites file and nogo areas are not supported   

Now, the « perfect » new Brouter-app should :
- support « nogo areas » as in the past (a very usefull feature !)
- cope with Google policies
- Support Android 11 users (and cope the new access methods of Android 11, as tested above)
- support Android versions lower 11

Is this your plan ?
(I think, it is possible, but it requires changes)
regards

**Comment by afischerdev** [comment]
> support « nogo areas » as in the past (a very usefull feature !)

It is supported - but in not the old way.

Old way was to read from external folder - in different ways - on three apps (OsmAnd, Locus, Orux).

New way is to read from internal folder (../Android/media/btools.routingapp/import). You can place a 'favourites.gpx' there and 'nogo*.gpx' in subfolder 'tracks'. This is open for all apps and BRouter only has to handle its own file access.

Old way with the three  CoordinateReader is still available and when **file access** is available they should work.

**Comment by EssBee59** [comment]
Thank for the information!
(The best is now for me to wait on a test version and discover the "new way")

On the other hand, "old way" should also work under Android 11, if as example OSMAND itself is installed in /Android/media.
Advantage is there, you have just to create or modify "nogo-favourites" in the Navigation app...
Ess Bee

**Comment by afischerdev** [comment]
@EssBee59 
> (The best is now for me to wait on a test version and discover the "new way")

I could again offer you a pre version with other signing data, if you like.

> On the other hand, "old way" should also work under Android 11

Will not work. When OsmAnd is in media as well, it still is an external folder for BRouter and need special handling.

**Comment by devemux86** [comment]
> for the Play Store a bundle *.aab file is needed for publishing

New apps are required to publish with the Android App Bundle on Google Play.

Updates to existing apps can continue with APK publishing format.

> Is it still useful to build the "Api10" version for direct installs without security restrictions?

This can be useful on Android 11+ so users can select a base dir where they can modify BRouter profiles.

**Comment by EssBee59** [comment]
> Will not work. When OsmAnd is in media as well, it still is an external folder for BRouter and need special handling.

Yes, to access /Android/media/net.osmand, brouter will need the old "storage access": But you can grant this permission manually to every app in the settings
But the "new way" you created is possibly better...
Not urgent, but if you can send a test version, I promise to test!

**Comment by afischerdev** [comment]
@devemux86 
I thought I read the November in the timeline for an updating ultimate. But could not find it again. 
Would be fine if APK will go on. Makes it easier.

> This can be useful on Android 11+ so users can select a base dir where they can modify BRouter profiles.

The .../Android/media/... is accessible for  other apps to edit the profiles. I use e.g. Total Commander.

@EssBee59 
> Yes, to access /Android/media/net.osmand, brouter will need the old "storage access": But you can grant this permission manually to every app in the settings

You are right, 'storage access' is possible, but it covers only the OsmOnd behaviors, gpx files can be read sequentially. But in Locus and Orux  a database format is used. In the past I didn't find a way to read sqlite databases outside the app internal storage.

So my focus still is to keep BRouter simple. That sounds may be a little lordly because it means all other apps have to move to the new features. But this could be used by any app and not only three.

**Comment by afischerdev** [comment]
@EssBee59 
after the merge #328 [here ](https://www.dropbox.com/s/mp5hece83bc36q5/brouter-1.6.1.zip?dl=0) the result zip file with both apks. Api 10 variante is not tested on api 10 hardware (I own only a Android 4.3 phone).

**Comment by EssBee59** [comment]
Hello afischerdev,

Sorry for the late answer, I was last week on a bike tour!
My tests:
- I could generate the brouterapp and brouter-jar without any problems using the sources in https://github.com/abrensch/brouter
- Gradle works perfectly, I also couls sign the app with my own key
- first tests under Android 11 are successfull, I just discovered a minor problem using the api10 version: osmand "nogo´s" did not work...
- but a simple change solved the problem:

as in this situationb dev = 30 AND targetsdk=19 I had to modify the code:
In BRouterService.java, line 206
And in BRouterView.java  line, 224
apply the following change
//      boolean canAccessSdCard =  deviceLevel < 23 || targetSdkVersion == 10;
      boolean canAccessSdCard =  deviceLevel < 23 || targetSdkVersion == 19;

Thank for providing the old api10 version, for me nogo is a very usefull feature!
Regards
Ess Bee

**Comment by afischerdev** [comment]
@EssBee59 
Great, I've changed the test to version 19. Old app was for api 10 only. It goes now from api 10 to 19.

**Comment by afischerdev** [comment]
Sorry for the delay I was on the way - testing is also a necessary thing. ;-)

So I will go on with the last things:
- icon: there was no reaction so no change
- BInstallerSizes: no reaction I'll update the factor
- variable values: no reaction I'll add it
- new version number

Still pending is the git workflow yml script for gradle

**Comment by afischerdev** [comment]
@zod 
Super, thanks for the new yml file. I've tested it and it works as expected. 
I'm now looking for the publish logic. Not sure if I need a publish section in each gradle file. Looks like.

Milestone seems a great idea, I've never used it before. But it would be eliminate the summaries from the list.

**Comment by zod** [comment]
I've started working on the publishing in https://github.com/zod/brouter/tree/gradle-publish. Currently it's partially uploading the first package but then it fails.

**Comment by afischerdev** [comment]
@zod 
Great, but I think, there is also a change on gradle files recommended.
Please hang on a minute I will change two as a sample.

**Comment by afischerdev** [comment]
@zod 
I've changed two gradle files in my [fork](https://github.com/afischerdev/brouter/tree/test-and11).
May be you can load from there.
Please change <USERNAME>/<REPO> to your values

**Comment by afischerdev** [comment]
@zod 
Super, I've merged it.

But ... on the syncing my fork and its branches I pressed the wrong button. Didn't notice the wrong direction. 
I have to repair that.

**Comment by zod** [comment]
I have a fix for the permission handling on my branch [permission-handling](https://github.com/zod/brouter/tree/permission-handling) but it's on top of the other PRs which are pending so I don't want to add another PR.

**Comment by afischerdev** [comment]
Good, I did the same. onRequestPermissionsResult was missing. 

But I was waiting for response of other users to finish the question of the api start at 10 or 14 at #354.

**Comment by EssBee59** [comment]
Hello,

I retested the API19 version based on the current sources, it "nogo´s" as allready succesfully tested (see above).
Every thing runs as expected, but I found a UI which may be a bit "confusing":
==> In this new API19 version, the "nogo´s" are always considered (as result of my tests, or did I a mistake?) 
==> In the UI-setting "BRouter app" ...==> "Check nogo Selection:" the user still have the possibility to select the desired nogo´s
As this selection has no effect (nogo´s are consired) it is confusing!
![brouter_nogo_UI](https://user-images.githubusercontent.com/50321352/144989836-1a14d983-c411-40ea-9f65-3025c7ec9f62.jpg)
Question:
Is it possible to optimize this point?
-The UI should remain, as it shows the curent nogo´s, proving that evrything about nogos is well configured  
- A comment indicating the NOUSE should reach
- Better soltution were to deactivate the selection-buttons
- A "nice to have" solution were:
Create a new button "Nogo Y/N" in this UI, then implement the optional nogo consideration in the app (BRouterService.java)

My use case for that:
I created nogo´s on very dangerous highways in my region
Depending on the day and the type of biking (alone using fastbike-verylowtraffic OR in groups of bikers in the week end) I liked to consider or not the nogo´s.

**Comment by afischerdev** [comment]
@EssBee59 
Thanks for the report. I'll have a look at it. 
What I see at the moment is:
NoGo reference is stored at 'modes/serviceconfig.dat' when selected and 'booked' for a profile.
But not used when nogo is loaded.
And this logic is only used when the 'old' mode is used for routing. With remote profile or call profile by name this will not work.
I think a new logic usable in all call variantes could be a nice new feature. 
On the other hand when you only store the usage of a nogo file there is no depending information for a profile.

**Comment by EssBee59** [comment]
Hello,

Yes, I had a look at the code!
My thoughts are only for "confort" and a change is not urgent!
But now I had a question: Waht do you mean with "or call profile by name"?
As now I have to implement my own code for this feature in BRouter and Osmand, I think, this is a future feature??
(I can send you my code by need)

Regards

**Comment by afischerdev** [comment]
@EssBee59 
We have three entries on BRouter interface:

```
    //  "v"-->[motorcar|bicycle|foot] mixed with "fast"
    //  "remoteProfile"--> (String), net-content of a profile. If remoteProfile != null, v+fast are ignored
    //  "profile"         = profile file name without .brf 

```
All of them have to get a nogo handling.
I understand, with 'nogo' parameter you setup 'only' use the nogos or not.
My intention was: e.g. you have 6 nogo files and 3 of them are selected. So on routing only selected should be used.

**Comment by afischerdev** [comment]
And more:
You have your nogo files bind to the profiles like 'old' logic for parameter 'v' does.
Which means you have 6 nogo files and bind 3 to bike, 2 of them to foot and one is bind to motorcar.

**Comment by afischerdev** [comment]
@EssBee59 
I add a test app at [and11_nogo](https://github.com/afischerdev/brouter/tree/and11_nogo).
The service works with profile and associated nogo files when in 'old' mode with parameter 'v'.
Otherwise it uses only selected nogo file. A file 'nogo votes.dat' is written to the modes folder.

**Comment by NorbertKre** [comment]
Today I installed BRouter new on S20, while updating from S9 on Android 10. After missing tiles after downloading and missing servercon fig.dat I found the release notes ;-)
But, while do we have the file structure completely  twice? And no link inside the folders to Android/media? Should be cleaned up. 

Another idea/wish: would be very helpful, if the Server-Config would show on start the actual Config. I use Auto for Racing Bike, Bike for Trekking-Bike and Pedestrian for Mountainbike. So, if we had a tool, which reads an prefered directory and allows mapping, would be helpful while routing in tools like OruxMaps.

Thanks Norbert

**Comment by afischerdev** [comment]
@NorbertKre 
> Today I installed BRouter new on S20, while updating from S9 on Android 10. After missing tiles after downloading and missing servercon fig.dat I found the release notes ;-)
But, while do we have the file structure completely twice? And no link inside the folders to Android/media? Should be cleaned up.

I guess you mean the same issue reported at #379.
Is that correct?

**Comment by NorbertKre** [comment]
Yes, didn't know about that. New in GitHub and this forum. 312 was the "link" from Arndt.. 

Regards Norbert 

⁣BlueMail for Android herunterladen ​

Am 26. Dez. 2021, 10:48, um 10:48, afischerdev ***@***.***> schrieb:
***@***.***

**Comment by afischerdev** [comment]
Closed with new release 1.6.3
